server/sonar-server/src/test/java/org/sonar/server/qualityprofile/CachingRuleActivatorTest.java

```diff
@@ -0,0 +1,142 @@
+    assertThat(underTest.getChildren(dbSession, "no_children"))
+        .isEmpty();
+    assertThat(underTest.getChildren(dbSession, "no_children"))
+        .isEmpty();
+    verify(qualityProfileDao, times(1)).selectChildren(eq(dbSession), anyString());
+  }
+
+  @Test
+  public void getChildren_caches_that_sq_has_on_children_and_cache_that_child_has_no_children() {
+    when(qualityProfileDao.selectChildren(dbSession, "0"))
+        .thenReturn(singletonList(dto("1")))
+        .thenThrow(new IllegalStateException("selectChildren should be called only once"));
+
+    assertThat(underTest.getChildren(dbSession, "0"))
+        .extracting(QualityProfileDto::getKey)
+        .containsExactly("1");
+    assertThat(underTest.getChildren(dbSession, "0"))
+        .extracting(QualityProfileDto::getKey)
+        .containsExactly("1");
+    assertThat(underTest.getChildren(dbSession, "0"))
+        .extracting(QualityProfileDto::getKey)
+        .containsExactly("1");
+    assertThat(underTest.getChildren(dbSession, "1"))
+        .isEmpty();
+    assertThat(underTest.getChildren(dbSession, "1"))
+        .isEmpty();
+    assertThat(underTest.getChildren(dbSession, "1"))
+        .isEmpty();
+    verify(qualityProfileDao, times(1)).selectChildren(eq(dbSession), anyString());
+  }
+
+  @Test
+  public void getChildren_caches_that_sq_has_children_and_recursively() {
+    when(qualityProfileDao.selectChildren(dbSession, "0"))
+      .thenReturn(asList(dto("1"), dto("2"), dto("3"), dto("4")))
+      .thenThrow(new IllegalStateException("selectChildren should be called only once"));
+
+    assertThat(underTest.getChildren(dbSession, "0"))
+      .extracting(QualityProfileDto::getKey)
+      .containsExactly("1", "2", "3", "4");
+    assertThat(underTest.getChildren(dbSession, "0"))
+      .extracting(QualityProfileDto::getKey)
+      .containsExactly("1", "2", "3", "4");
+    assertThat(underTest.getChildren(dbSession, "1"))
+      .extracting(QualityProfileDto::getKey)
+      .containsExactly("2", "3", "4");
+    assertThat(underTest.getChildren(dbSession, "1"))
+      .extracting(QualityProfileDto::getKey)
+      .containsExactly("2", "3", "4");
+    assertThat(underTest.getChildren(dbSession, "2"))
+      .extracting(QualityProfileDto::getKey)
+      .containsExactly("3", "4");
+    assertThat(underTest.getChildren(dbSession, "2"))
+      .extracting(QualityProfileDto::getKey)
+      .containsExactly("3", "4");
+    assertThat(underTest.getChildren(dbSession, "3"))
+      .extracting(QualityProfileDto::getKey)
+      .containsExactly("4");
+    assertThat(underTest.getChildren(dbSession, "3"))
+      .extracting(QualityProfileDto::getKey)
+      .containsExactly("4");
+    assertThat(underTest.getChildren(dbSession, "4"))
+      .isEmpty();
+    assertThat(underTest.getChildren(dbSession, "4"))
+      .isEmpty();
+
+    verify(qualityProfileDao, times(1)).selectChildren(eq(dbSession), anyString());
+  }
+
+  private QualityProfileDto dto(String key) {
```

Comment:
I wanted to override `toString()` for more convenient assert errors, so used this while I was at it.
Could also work with both override and subsequent call to `setKey`